File: backend/app/games/pong/state.py

import random
from dataclasses import dataclass, field

from .constants import INITIAL_VELOCITY_X, INITIAL_VELOCITY_Z, Side

@dataclass
class SimState:

    ball_x: float = 0.0
    ball_z: float = 0.0
    velocity_x: float = 0.0
    velocity_z: float = 0.0
    left_paddle_offset: float = 0.0
    right_paddle_offset: float = 0.0
    left_paddle_velocity: float = 0.0
    right_paddle_velocity: float = 0.0
    last_left_paddle_input_at: float = 0.0
    last_right_paddle_input_at: float = 0.0
    score_left: int = 0
    score_right: int = 0
    winner: Side | None = field(default=None)

def create_initial_sim_state() -> SimState:
    return SimState(
        velocity_x=INITIAL_VELOCITY_X,
        velocity_z=INITIAL_VELOCITY_Z * _random_velocity_z_sign(),
    )

def _random_velocity_z_sign() -> int:
    return 1 if random.random() > 0.5 else -1

def reset_ball(state: SimState, toward_right: bool) -> None:
    state.ball_x = 0
    state.ball_z = 0
    state.velocity_x = INITIAL_VELOCITY_X if toward_right else -INITIAL_VELOCITY_X
    state.velocity_z = INITIAL_VELOCITY_Z * _random_velocity_z_sign()

def stop_ball(state: SimState) -> None:
    state.ball_x = 0
    state.ball_z = 0
    state.velocity_x = 0
    state.velocity_z = 0

def is_match_finished(state: SimState) -> bool:
    return state.winner is not None

def to_snapshot(state: SimState, timestamp: float) -> dict:
    return {
        "leftPaddleOffset": state.left_paddle_offset,
        "rightPaddleOffset": state.right_paddle_offset,
        "ball": {"x": state.ball_x, "z": state.ball_z},
        "score": {"left": state.score_left, "right": state.score_right},
        "winner": state.winner,
        "timestamp": timestamp,
    }

File: frontend/src/mocks/pong/simulation/state.ts
import type { GameSnapshot } from "../../../games/pong/protocol/types";
import {
  INITIAL_VELOCITY_X,
  INITIAL_VELOCITY_Z,
} from "./constants";
import type { SimState } from "./types";

/** Sorteia o sinal da velocidade inicial no eixo Z (largura) (+1 ou -1). */
function randomVelocityZSign(): number {
  return Math.random() > 0.5 ? 1 : -1;
}

/** Cria o estado inicial de uma partida. */
export function createInitialSimState(): SimState {
  return {
    ballX: 0,
    ballZ: 0,
    velocityX: INITIAL_VELOCITY_X,
    velocityZ: INITIAL_VELOCITY_Z,
    leftPaddleOffset: 0,
    rightPaddleOffset: 0,
    leftPaddleVelocity: 0,
    rightPaddleVelocity: 0,
    lastLeftPaddleInputAt: 0,
    lastRightPaddleInputAt: 0,
    scoreLeft: 0,
    scoreRight: 0,
    winner: null,
  };
}

/**
 * Reposiciona a bola no centro e define a direção do saque.
 *
 * @param state - Estado da simulação
 * @param towardRight - Se `true`, a bola sai no sentido +X; caso contrário, −X
 */
export function resetBall(state: SimState, towardRight: boolean): void {
  state.ballX = 0;
  state.ballZ = 0;
  state.velocityX = towardRight ? INITIAL_VELOCITY_X : -INITIAL_VELOCITY_X;
  state.velocityZ = INITIAL_VELOCITY_Z * randomVelocityZSign();
}

/** Para a bola e a posiciona no centro da mesa. */
export function stopBall(state: SimState): void {
  state.ballX = 0;
  state.ballZ = 0;
  state.velocityX = 0;
  state.velocityZ = 0;
}

/** Indica se a partida já possui um vencedor. */
export function isMatchFinished(state: SimState): boolean {
  return state.winner !== null;
}

/**
 * Converte o estado interno da simulação para o GameSnapshot do protocolo.
 *
 * @param state - Estado da simulação
 * @param timestamp - Timestamp do frame, em milissegundos
 */
export function toSnapshot(state: SimState, timestamp: number): GameSnapshot {
  return {
    leftPaddleOffset: state.leftPaddleOffset,
    rightPaddleOffset: state.rightPaddleOffset,
    ball: { x: state.ballX, z: state.ballZ },
    score: { left: state.scoreLeft, right: state.scoreRight },
    winner: state.winner,
    timestamp,
  };
}
